core/src/test/scala/org/apache/spark/sql/delta/MergeIntoSQLSuite.scala

```diff
@@ -82,7 +82,8 @@ class MergeIntoSQLSuite extends MergeIntoSuiteBase  with DeltaSQLCommandTest
         update = "key2 = 20 + src.key3, value = 20 + src.value",
         insert = "(key2, value) VALUES (src.key3 - 10, src.value + 10)")
 
-      sql(cte + merge)
+//      print(sql(cte + merge).explain(true))
+      checkAnswer(sql(cte + merge), Row(2, 1, 0, 1))
```

Comment:
while printing this sql(cte +merge) it is giving fine output 

```

+-----------------------+----------------------+----------------------+-----------------------+
|number of affected rows|number of updated rows|number of deleted rows|number of inserted rows|
+-----------------------+----------------------+----------------------+-----------------------+
|                      2|                     1|                     0|                      1|
+-----------------------+----------------------+----------------------+-----------------------+
```

But when i am running this checkAnswer i am getting this error . 
Has anyone ever faced this issue ?
```
AttributeSet(key3#573, key2#574, value#577) was not empty The analyzed logical plan has missing inputs:
MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-8cc2e21e-e23a-4ac4-9055-0f1ef9d53058], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

ScalaTestFailureLocation: org.apache.spark.sql.QueryTest at (QueryTest.scala:225)
org.scalatest.exceptions.TestFailedException: AttributeSet(key3#573, key2#574, value#577) was not empty The analyzed logical plan has missing inputs:
MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-8cc2e21e-e23a-4ac4-9055-0f1ef9d53058], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

```

plans For DataFrame 

```
      == Parsed Logical Plan ==
CTE [cte1]
:  +- 'SubqueryAlias cte1
:     +- 'Project [('key1 + 2) AS key3#573, 'value]
:        +- 'UnresolvedRelation [source], [], false
+- 'MergeIntoTable ('src.key3 = 'target.key2), [updateaction(None, assignment('key2, (20 + 'src.key3)), assignment('value, (20 + 'src.value)))], [insertaction(None, assignment('key2, ('src.key3 - 10)), assignment('value, ('src.value + 10)))]
   :- 'SubqueryAlias target
   :  +- 'UnresolvedRelation [delta, /private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-0cfff9af-311a-4a19-b4e9-975bec5549f1], [], false
   +- 'SubqueryAlias src
      +- 'UnresolvedRelation [cte1], [], false

== Analyzed Logical Plan ==
number of affected rows: bigint, number of updated rows: bigint, number of deleted rows: bigint, number of inserted rows: bigint
MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-0cfff9af-311a-4a19-b4e9-975bec5549f1], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

== Optimized Logical Plan ==
CommandResult [number of affected rows#580L, number of updated rows#581L, number of deleted rows#582L, number of inserted rows#583L], Execute MergeIntoCommand, [[2,1,0,1]]
   +- MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-0cfff9af-311a-4a19-b4e9-975bec5549f1], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

== Physical Plan ==
CommandResult [number of affected rows#580L, number of updated rows#581L, number of deleted rows#582L, number of inserted rows#583L]
   +- Execute MergeIntoCommand
         +- MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-0cfff9af-311a-4a19-b4e9-975bec5549f1], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

```

Reply:
#1240 is also giving the same error .

Reply:
can you show the full stacktrace please?

Reply:
```
AttributeSet(key3#573, key2#574, value#577) was not empty The analyzed logical plan has missing inputs:
MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-e25728d6-d7fb-4312-8804-b86837ca838c], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

ScalaTestFailureLocation: org.apache.spark.sql.QueryTest at (QueryTest.scala:225)
org.scalatest.exceptions.TestFailedException: AttributeSet(key3#573, key2#574, value#577) was not empty The analyzed logical plan has missing inputs:
MergeIntoCommand SubqueryAlias src, SubqueryAlias target, Delta[version=1, ... :/private/var/folders/8m/10wm3v9s6nzbm6pjk2vldg6r0000gn/T/spark-e25728d6-d7fb-4312-8804-b86837ca838c], (key3#573 = key2#574), [Update [actions: [`key2` = (20 + key3#573), `value` = (20 + value#577)]]], [Insert [actions: [`key2` = (key3#573 - 10), `value` = (value#577 + 10)]]], StructType(StructField(key2,IntegerType,true),StructField(value,IntegerType,true))

	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
	at org.apache.spark.sql.QueryTest.assertEmptyMissingInput(QueryTest.scala:225)
	at org.apache.spark.sql.QueryTest.checkAnswer(QueryTest.scala:149)
	at org.apache.spark.sql.QueryTest.checkAnswer(QueryTest.scala:155)
	at org.apache.spark.sql.delta.MergeIntoSQLSuite.$anonfun$new$2(MergeIntoSQLSuite.scala:86)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.sql.test.SQLTestUtilsBase.withTable(SQLTestUtils.scala:306)
	at org.apache.spark.sql.test.SQLTestUtilsBase.withTable$(SQLTestUtils.scala:304)
	at org.apache.spark.sql.delta.MergeIntoSuiteBase.withTable(MergeIntoSuiteBase.scala:38)
	at org.apache.spark.sql.delta.MergeIntoSQLSuite.$anonfun$new$1(MergeIntoSQLSuite.scala:73)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:203)
	at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:64)
	at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
	at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
	at org.apache.spark.SparkFunSuite.runTest(SparkFunSuite.scala:64)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTests$1(AnyFunSuiteLike.scala:269)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests(AnyFunSuiteLike.scala:269)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTests$(AnyFunSuiteLike.scala:268)
	at org.scalatest.funsuite.AnyFunSuite.runTests(AnyFunSuite.scala:1563)
	at org.scalatest.Suite.run(Suite.scala:1112)
	at org.scalatest.Suite.run$(Suite.scala:1094)
	at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1563)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
	at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
	at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:64)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:64)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1322)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1316)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1316)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:993)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:971)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1482)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:971)
	at org.scalatest.tools.Runner$.run(Runner.scala:798)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2or3(ScalaTestRunner.java:38)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:25)
```

Reply:
I spent some time digging. This is strange.

Reply:
It is also happening with the `UpdateCommand` (I was attempting to work on #1321).

Reply:
@Silverlight42 yep this one works for me . @scottsand-db can i use QueryTest.checkAnswer one instead of checkAnswer ?

Reply:
not sure if this help or not  . 
checkAnswer definition is 
```
protected def checkAnswer(df: => DataFrame, expectedAnswer: Seq[Row]): Unit = {
    val analyzedDF = try df catch {
      case ae: AnalysisException =>
        if (ae.plan.isDefined) {
          fail(
            s"""
               |Failed to analyze query: $ae
               |${ae.plan.get}
               |
               |${stackTraceToString(ae)}
               |""".stripMargin)
        } else {
          throw ae
        }
    }

    assertEmptyMissingInput(analyzedDF)

    QueryTest.checkAnswer(analyzedDF, expectedAnswer)
  }
```
so when we are running QueryTest.checkAnswer directly we are skipping assertEmptyMissingInput (the error generating code )

Reply:
Hi @sherlockbeard and @Silverlight42 . Using `QueryTest.checkAnswer(df, expectedAnswer)` is fine here.

`checkAnswer` is not supposed to be called with actual commands (like UPDATE, MERGE, DESCRIBE HISTORY, etc). In fact, if you added some sort of projection ... like
```scala
checkAnswer(sql(cte + merge).select("number of affected rows"), Row(2))
```

then this would actually pass/work fine!

So, nothing wrong with your code :) Just use `QueryTest.checkAnswer(analyzedDF, expectedAnswer)` and skip the `assertEmptyMissingInput` call.